codechain/primitives/AssetAddress.py

```diff
@@ -0,0 +1,110 @@
+import re
+from typing import List
+from typing import NamedTuple
+from typing import Tuple
+from typing import Union
+
+from ..crypto import bech32_decode
+from ..crypto import bech32_encode
+from .HexString import H160
+
+
+class MultiSig(NamedTuple):
+    n: int
+    m: int
+    pubkeys: List[H160]
+
+
+class AssetAddress:
+    def __init__(
+        self, address_type: int, payload: Union[H160, str, MultiSig], address: str
+    ):
+        self.address_type = address_type
+
+        if isinstance(payload, (H160, str)):
+            self.payload = H160(payload)
+        else:
+            n, m, pubkeys = payload
+            self.payload = (n, m, list(map(lambda p: H160(p), pubkeys)))
+        self.value = address
+
+    @classmethod
+    def from_type_and_payload(
+        cls, network_type: int, payload: Union[H160, str, MultiSig], **kwargs
+    ):
+        if "network_id" in kwargs:
+            network_id = kwargs.get("network_id")
+        else:
+            network_id = "cc"
+        if "version" in kwargs:
+            version = kwargs.get("version")
+        else:
+            version = 1
+
+        if version != 1:
+            raise ValueError(f"Unsupported version for asset address: {version}")
+        if network_type < 0x00 or network_type > 0x03:
+            raise ValueError(f"Unsupported type for asset address: {network_type}")
+
+        address = bech32_encode(
+            network_id + "a", bytes([version, network_type]) + encode_payload(payload)
+        )
+
+        return cls(network_type, payload, address)
+
+    @classmethod
+    def from_string(cls, address: str):
+        if address[2] != "a":
+            ValueError(f"The prefix is unknown for asset address: {address}")
+
+        byte = bech32_decode(address[:3], address)
+        version = byte[0]
+
+        if version != 1:
+            ValueError(f"Unsupported version for asset address: {address}")
+
+        address_type = byte[1]
+
+        if address_type < 0x0 or address_type > 0x3:
+            ValueError(f"Unsupported type for asset address: {address}")
+
+        if address_type < 0x3:
+            payload = H160(byte[2:])
+            return cls(address_type, payload, address)
+        else:
+            n = byte[2]
+            m = byte[3]
+            payload = byte[4:]
+            pubkeys = []
+            for i in range(0, len(payload), 20):
+                pubkeys.append(H160(payload[i : i + 20]))
+
+            return cls(address_type, MultiSig(n, m, pubkeys), address)
+
+    @classmethod
+    def check(cls, address):
+        return True if isinstance(address, cls) else cls.check_string(address)
+
+    @classmethod
+    def check_string(cls, address: str):
+        return None is not re.match(
+            "^.{2}a[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{42}$", address
+        )
+
+    @classmethod
+    def ensure(cls, address):
+        return address if isinstance(address, cls) else cls.from_string(address)
+
+    def to_string(self) -> str:
```

Comment:
We can use `def __str__(self)` for this since it doesn't require any formating parameters.

Reply:
Thanks